File: Backend/StockFlow.Application/Services/CategoriesService.cs
using StockFlow.Application.Helpers;
using StockFlow.Application.Interfaces;
using StockFlow.Domain.Entities;

namespace StockFlow.Application.Services;

public class CategoriesService
{
    private readonly ICategoryRepository _categoryRepository;

    public CategoriesService(ICategoryRepository categoryRepository)
    {
        _categoryRepository = categoryRepository;
    }

    public async Task<IEnumerable<Categories>> GetCategories()
    {
        return await _categoryRepository.GetCategories();
    }

    public async Task<Categories> GetCategoryById(Guid categoryId)
    {
        var category = await _categoryRepository.GetCategoryById(categoryId);

        if (category == null) ErrorHelper.ThrowNotFoundException("Category not found!");

        return category;
    }

    public async Task<Categories> CreateCategory(Categories category)
    {
        return await _categoryRepository.CreateCategory(category);
    }

    public async Task<Categories> UpdateCategory(Guid categoryId, Categories updateCategory)
    {
        var currentCategory = await GetCategoryById(categoryId);

        currentCategory.Name = updateCategory.Name;

        await _categoryRepository.UpdateCategory(currentCategory);
        return currentCategory;
    }

    public async Task DeleteCategory(Guid categoryId)
    {
        await _categoryRepository.DeleteCategory(categoryId);
    }
}

File: Frontend/stockflow.client/src/services/categoriesService.ts
import { ICategory } from "../@types/category";
import apiRequest from "./helpers/apiService";

export const GetCategories = async () => apiRequest("GET", "categories", undefined, true);

export const GetCategoryById = async (categoryId: string) => apiRequest("GET", `categories/${categoryId}`, undefined, true);

export const CreateCategory = async (newCategory: ICategory) => apiRequest("POST", "categories", newCategory, true);

export const UpdateCategory = async (categoryId: string, updateCategory: Partial<ICategory>) =>
    apiRequest("PUT", `categories/${categoryId}`, updateCategory, true);

export const DeleteCategory = async (categoryId: string) => apiRequest("DELETE", `categories/${categoryId}`, undefined, true);